drivers/built_for_zero_report/app/models/built_for_zero_report/calculator.rb

```diff
@@ -0,0 +1,151 @@
+    end
+
+    # @return [SourceDataHash] clients in cohort who returned to homelessness from housing in the reporting period
+    def returned_from_housing
+      source_data.select { |_, v| v[:change] == 'create' && v[:reason] == 'Returned from housing' }
+    end
+
+    # @return [SourceDataHash] clients who returned to the cohort after a period of inactivity in the reporting period
+    def returned_from_inactivity
+      source_data.select { |_, v| v[:change] == 'create' && v[:reason] == 'Returned from inactive' }
+    end
+
+    # Roll up of a cohort change history to reflect most recent change for each client on a specified date.
+    # @return [scope] GrdaWarehouse::Hud::Clients joined with GrdaWarehouse::CohortClientChange
+    private def source_scope
+      mrc_t = Arel::Table.new(:most_recent_changes)
+      join = c_c_change_t.join(mrc_t).on(c_c_change_t[:id].eq(mrc_t[:current_id]))
+
+      GrdaWarehouse::Hud::Client.
+        with(
+          most_recent_changes:
+            GrdaWarehouse::CohortClientChange.
+              define_window(:cohort_client_id_by_change_date).partition_by(c_c_change_t[:cohort_client_id], order_by: { c_c_change_t[:changed_at] => :desc }).
+              select_window(:first_value, c_c_change_t[:id], over: :cohort_client_id_by_change_date, as: :current_id).
+              where(cohort_id: @cohort_id).
+              where(c_c_change_t[:changed_at].lteq(@end_date)),
+        ).
+        joins(cohort_clients: :cohort_client_changes).
+        joins(join.join_sources)
+    end
+
+    # Get the source data from specified cohort in the reporting period.
+    # @return [Hash{Integer => Hash}] Hash of destination client IDs to source data hashes
+    private def source_data
+      @source_data ||= begin
+        fields = {
+          client_id: c_t[:id], # Client ID must be first as it is used as the hash key
+          first_name: c_t[:FirstName],
+          last_name: c_t[:LastName],
+          change: c_c_change_t[:change],
+          reason: c_c_change_t[:reason],
+          changed_at: c_c_change_t[:changed_at],
+        }
+        data = source_scope.
+          pluck(*([c_t[:id]] + fields.values)).
+          group_by(&:first).
+          transform_values { |arr| fields.keys.zip(arr.flatten).to_h }.
+          reject { |_, v| v[:change] == 'destroy' && v[:changed_at] < @start_date } # Only include exits in the reporting period
+
+        data = data.keep_if { |k, _| k.in?(@client_ids) } if @client_ids.present?
+        data
+      end
+    end
+
+    # Extend the source data hash from the housed clients to include LOT data
+    # @return [Hash{Integer => Hash}] Hash of destination client IDs to extended source data hashes
+    def lot_to_housing
+      @lot_to_housing ||= begin
+        housed_clients = housed
+        first_enrollment_dates = GrdaWarehouse::ServiceHistoryEnrollment.
+          first_date.
+          where(client_id: housed_clients.keys).pluck(:client_id, :first_date_in_program).
```

Comment:
minimum of Enrollment [DateToStreetESSH, EntryDate] where the enrollment 3.917 Prior Living Situation is in `HUD.homeless_situations`